### PR TITLE
Jax with CUDA support

### PR DESCRIPTION
**Pull request by rabernat**
See https://github.com/pangeo-data/pangeo-docker-images/pull/325#issuecomment-1128068097

When I tried to condalock this, I got

```
generating lockfile for linux-64
Failed to parse json, Expecting value: line 1 column 1 (char 0)
Could not lock the environment for platform linux-64
    Command: ['/opt/miniconda3/envs/condalock/bin/mamba', 'create', '--prefix', '/var/folders/n8/63q49ms55wxcj_gfbtykwp5r0000gn/T/tmpxbvhi8ui/prefix', '--dry-run', '--json', '--override-channels', '--channel', 'ngam', '--channel', 'conda-forge', 'xarray_leaflet', 'geopandas', 'pystac', 'xesmf', 'xcape', 'nc-time-axis', 'adlfs', 'xrft', 'earthdata', 'rechunker', 'param!=1.11.0', 'xarray-spatial', 'zarr', 'scikit-image', 'nb_conda_kernels', 'timezonefinder', 'xskillscore', 'jupyter-resource-usage', 'pandas', 'satpy', 'ipywidgets', 'scikit-learn', 'matplotlib-base', 'panel', 'numcodecs', 'xarray', 'h5netcdf', 'anaconda::libopenblas', 'netcdf4', 'python=3.9*', 'erddapy', 'jaxlib==*=*cuda*', 'xarrayutils', 'nbstripout', 'pangeo-notebook=2022.05.10', 'pip=20', 'pyarrow', 'scipy', 'intake-xarray', 'parcels', 'ipytree', 'rasterio', 'rioxarray', 'awscli', 'cartopy >= 0.20.0', 'cudatoolkit=10', 'pystac-client', 'xpublish', 'geocube', 'pygeos', 'kerchunk', 'lz4', 'intake-esm', 'eofs', 'python-graphviz', 'geopy', 'gh', 'gh-scoped-creds', 'git-lfs', 'sparse', 'argopy', 'jax=0.3.0', 'intake-geopandas', 'xgcm', 'metpy', 'geoviews-core', 'intake-stac', 'rio-cogeo', 's3fs>0.5', 'pop-tools', 'bottleneck', 'esmpy', 'h5py', 'fsspec', 'jupyter-panel-proxy', 'dask-ml', 'hvplot', 'lxml', 'seaborn', 'ciso', 'stackstac', 'gsw', 'boto3', 'gcsfs', 'pycamhd', 'tensorflow=2.7.0', 'cfgrib', 'tiledb-py', 'nomkl', 'datashader', 'intake', 'ipyleaflet', 'python-gist', 'xmitgcm', 'fastjmd95', 'python-blosc', 'pydap', 'gcm_filters', 'prefect', 'cmip6_preprocessing', 'xhistogram', 'descartes', 'numpy', 'holoviews']
    STDOUT:
Encountered problems while solving.
Problem: nothing provides __glibc >=2.17 needed by jaxlib-0.3.7-cuda112py39h9c5d0a6_0
```

cc @ngam

### CODE REVIEW

**Comment by ngam** [comment]
> See [#325 (comment)](https://github.com/pangeo-data/pangeo-docker-images/pull/325#issuecomment-1128068097)
> 
> When I tried to condalock this, I got
> 
> ```
> generating lockfile for linux-64
> Failed to parse json, Expecting value: line 1 column 1 (char 0)
> Could not lock the environment for platform linux-64
>     Command: ['/opt/miniconda3/envs/condalock/bin/mamba', 'create', '--prefix', '/var/folders/n8/63q49ms55wxcj_gfbtykwp5r0000gn/T/tmpxbvhi8ui/prefix', '--dry-run', '--json', '--override-channels', '--channel', 'ngam', '--channel', 'conda-forge', 'xarray_leaflet', 'geopandas', 'pystac', 'xesmf', 'xcape', 'nc-time-axis', 'adlfs', 'xrft', 'earthdata', 'rechunker', 'param!=1.11.0', 'xarray-spatial', 'zarr', 'scikit-image', 'nb_conda_kernels', 'timezonefinder', 'xskillscore', 'jupyter-resource-usage', 'pandas', 'satpy', 'ipywidgets', 'scikit-learn', 'matplotlib-base', 'panel', 'numcodecs', 'xarray', 'h5netcdf', 'anaconda::libopenblas', 'netcdf4', 'python=3.9*', 'erddapy', 'jaxlib==*=*cuda*', 'xarrayutils', 'nbstripout', 'pangeo-notebook=2022.05.10', 'pip=20', 'pyarrow', 'scipy', 'intake-xarray', 'parcels', 'ipytree', 'rasterio', 'rioxarray', 'awscli', 'cartopy >= 0.20.0', 'cudatoolkit=10', 'pystac-client', 'xpublish', 'geocube', 'pygeos', 'kerchunk', 'lz4', 'intake-esm', 'eofs', 'python-graphviz', 'geopy', 'gh', 'gh-scoped-creds', 'git-lfs', 'sparse', 'argopy', 'jax=0.3.0', 'intake-geopandas', 'xgcm', 'metpy', 'geoviews-core', 'intake-stac', 'rio-cogeo', 's3fs>0.5', 'pop-tools', 'bottleneck', 'esmpy', 'h5py', 'fsspec', 'jupyter-panel-proxy', 'dask-ml', 'hvplot', 'lxml', 'seaborn', 'ciso', 'stackstac', 'gsw', 'boto3', 'gcsfs', 'pycamhd', 'tensorflow=2.7.0', 'cfgrib', 'tiledb-py', 'nomkl', 'datashader', 'intake', 'ipyleaflet', 'python-gist', 'xmitgcm', 'fastjmd95', 'python-blosc', 'pydap', 'gcm_filters', 'prefect', 'cmip6_preprocessing', 'xhistogram', 'descartes', 'numpy', 'holoviews']
>     STDOUT:
> Encountered problems while solving.
> Problem: nothing provides __glibc >=2.17 needed by jaxlib-0.3.7-cuda112py39h9c5d0a6_0
> ```
> 
> cc @ngam

This is likely very misleading 😅 (It's been such a pain...)

The problem is likely a combo of needing cudatoolkit 11.2+ and maybe an issue with jax. Let's unpin jax itself (generally jax's version should be >= jaxlib's version)

File: ml-notebook/environment.yml [hunk 1]
```diff
@@ -2,9 +2,10 @@
 # by conda-lock. Only put things here that are not already in pangeo-notebook
 name: pangeo
 channels:
+ - ngam
  - conda-forge
 dependencies:
  - cudatoolkit=10
```
Comment (ngam) [inline ml-notebook/environment.yml:8]:
```suggestion
 - cudatoolkit=11
```

File: ml-notebook/environment.yml [hunk 2]
```diff
@@ -2,9 +2,10 @@
...
  - tensorflow=2.7.0
  - jax=0.3.0
```
Comment (ngam) [inline ml-notebook/environment.yml:10]:
```suggestion
 - jax=0.3
```

**Comment by rabernat** [comment]
Thanks for the suggestion @ngam! I tried those changes but unfortunately am still getting

```
Problem: nothing provides __glibc >=2.17 needed by jaxlib-0.3.7-cuda112py39h9c5d0a6_0
```

(I assume the [running action](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true) will show the same failure eventually.)

I have to step away from this for the rest of the day. But I really appreciate your help getting these complex packages working. It's such an important service to our community! 👏

**Comment by rabernat** [comment]
Actually the CI condalock error is even weirder:

```
  - nothing provides openssl >=1.1.1,<1.1.2.0a0 needed by python-3.7.1-h0371630_3
```

Why in the world would we be looking for python 3.7 here... 😕  And why would it not give the same results as my local env.

**Comment by ngam** [comment]
The error is slightly different from the action:

```
Run conda clean --all --yes --force-pkgs-dirs
[10](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:10)
Remove all contents from the following package caches?
[11](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:11)
Generating lockfile(s) for linux-64...
[12](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:12)
Failed to parse json, Expecting value: line 1 column 1 (char 0)
[13](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:13)
Could not lock the environment for platform linux-64
[14](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:14)
    Command: /usr/share/miniconda3/envs/condalock/bin/mamba create --prefix /tmp/tmpem7sc5a0/prefix --dry-run --json --override-channels --channel ngam --channel conda-forge --channel file:///home/runner/work/_temp/tmpmcwcjppv prefect xrft pygeos satpy bottleneck panel pangeo-notebook=2022.05.10 xarrayutils xpublish python-graphviz jupyter-resource-usage 'python=3.9*' geopandas metpy 'jaxlib==*=*cuda*' gh gcsfs pystac 's3fs>0.5' hvplot netcdf4 scipy xcape datashader fsspec holoviews h5netcdf rioxarray timezonefinder tiledb-py argopy cfgrib gcm_filters zarr eofs xarray_leaflet jupyter-panel-proxy intake-xarray rasterio rio-cogeo cudatoolkit=11 h5py awscli matplotlib-base parcels cmip6_preprocessing jax=0.3 xgcm pystac-client rechunker xarray nc-time-axis scikit-image xesmf nomkl ipyleaflet nbstripout erddapy python-blosc sparse stackstac kerchunk descartes pyarrow ciso intake xarray-spatial pip=20 intake-geopandas pandas geocube pycamhd lxml gh-scoped-creds fastjmd95 ipywidgets pop-tools geopy anaconda::libopenblas xskillscore xhistogram 'param!=1.11.0' numpy tensorflow=2.7.0 intake-stac lz4 'cartopy >= 0.20.0' python-gist dask-ml git-lfs pydap scikit-learn earthdata gsw esmpy boto3 nb_conda_kernels seaborn xmitgcm intake-esm adlfs ipytree geoviews-core numcodecs
[15](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:15)
    STDOUT:
[16](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:16)
Encountered problems while solving:
[17](https://github.com/pangeo-data/pangeo-docker-images/runs/6459476119?check_suite_focus=true#step:8:17)
  - nothing provides openssl >=1.1.1,<1.1.2.0a0 needed by python-3.7.1-h0371630_3
```

I am almost certain the `__glibc >=2.17` is not the issue here. Something else is

**Comment by scottyhq** [comment]
Not sure it will help, but I'd try without CI, perhaps with the latest conda-lock (need to merge https://github.com/pangeo-data/pangeo-docker-images/pull/314)

See also these docs that seem relevant, but I can't say i've used this feature or fully understand it:
https://github.com/conda-incubator/conda-lock#--virtual-package-spec

**Comment by ngam** [comment]
> Not sure it will help, but I'd try without CI, perhaps with the latest conda-lock (need to merge #314)
> 
> See also these docs that seem relevant, but I can't say i've used this feature or fully understand it: https://github.com/conda-incubator/conda-lock#--virtual-package-spec

I was gonna say that'd be the next thing for me to try locally, usually replacing:

```
mamba create -n jax39 python=3.9 jaxlib==*=*cuda* jax -c ngam -c conda-forge 
```

with

```
CONDA_OVERRIDE_CUDA="11.2" mamba create -n jax39 python=3.9 jaxlib==*=*cuda* jax -c ngam -c conda-forge
```

where `CONDA_OVERRIDE_CUDA` basically sets that `__cuda` package spec thingy. Though to be clear, it is a little all over the place how that `__cuda` works, so the details in this issue (and some associated controversy initiated by me...) https://github.com/conda-forge/tensorflow-feedstock/issues/208

**Comment by ngam** [comment]
This looks pretty much exactly like this https://github.com/conda-forge/tensorflow-feedstock/issues/224, but don't ask me how it crept in here...

**Pull request closed without merging**